shared/reducers/tracker.js

```diff
@@ -1,842 +1,399 @@
+    case TrackerGen.updateProofState: {
+      const {username} = action.payload
+      return updateUserState(state, username, state => {
+        if (!state) {
+          return state
+        }
+        const proofsGeneralState = Constants.overviewStateOfProofs(state.proofs)
+        const trackerMessage = Constants.deriveTrackerMessage(
+          username,
+          state.currentlyFollowing,
+          proofsGeneralState
+        )
+        const reason = trackerMessage || state.reason
+        return {
+          ...state,
+          changed: proofsGeneralState.anyChanged,
+          reason,
+          shouldFollow: Constants.deriveShouldFollow(proofsGeneralState),
+          trackerState: Constants.deriveSimpleProofState(state.eldestKidChanged, proofsGeneralState),
+        }
+      })
+    }
+    case TrackerGen.resetProofs:
+      return updateUserState(state, action.payload.username, state => ({
         ...state,
         proofs: [],
-      }
-
-    case Constants.setProofs:
-      if (!action.payload) {
-        return state
-      }
-
-      const identity: RPCTypes.Identity = action.payload.identity
-      return {
+      }))
+    case TrackerGen.setProofs: {
+      const {identity, username} = action.payload
+      return updateUserState(state, action.payload.username, state => ({
         ...state,
-        proofs: dedupeProofs([
-          ...state.proofs,
-          ...(identity.revokedDetails || []).map(rv => revokedProofToProof(rv)),
-          ...(identity.proofs || []).map(rp => remoteProofToProof(username, Constants.checking, rp.proof)),
+        proofs: Constants.dedupeProofs([
+          ...(state ? state.proofs : []),
+          ...(identity.revokedDetails || []).map(rv => Constants.revokedProofToProof(rv)),
+          ...(identity.proofs || [])
+            .map(rp => Constants.remoteProofToProof(username, Constants.checking, rp.proof)),
         ]),
-      }
-
-    case Constants.updatePGPKey: {
-      if (!action.payload) {
-        return state
-      }
-      const url = `https://keybase.io/${state.username}/sigchain`
-      const proof = {
-        color: 'green',
-        // TODO: We don't currently get the sigID so we can't link to the actual sigChain statement. See https://keybase.atlassian.net/browse/CORE-3529
-        humanUrl: url,
-        id: action.payload.kid,
-        isTracked: state.currentlyFollowing,
-        mTime: 0,
-        meta: null,
-        name: action.payload.fingerPrint,
-        profileUrl: url,
-        state: 'normal',
-        type: 'pgp',
-      }
-
-      return {
-        ...state,
-        proofs: dedupeProofs(state.proofs.concat([proof])),
-      }
+      }))
     }
+    case TrackerGen.updatePGPKey: {
+      const {pgpFingerprint, kid} = action.payload
+      const fingerPrint = Constants.bufferToNiceHexString(pgpFingerprint)
+      return updateUserState(state, action.payload.username, state => {
+        const url = `https://keybase.io/${state ? state.username : ''}/sigchain`
```

Comment:
this is from the old code, but if state is null then the url will be bogus. Perhaps we can use the username from the payload? (here and similar cases below)

Reply:
i think in practice this never happens so its not a big deal. really i think we'll convert this to immutable and enties and we have have stronger guarantees and never allow a null username